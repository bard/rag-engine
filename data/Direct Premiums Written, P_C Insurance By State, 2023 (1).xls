
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  </head>
  <body>
    <table>
    <thead><tr><th></th></tr></thead>    <tbody> <tr class="odd"><td><table>
<tr><td>Direct Premiums Written, P/C Insurance By State, 2023 (1)</td></tr>
<tr><td><p>($000)</p>
</td></tr>
<tr><td><table>
<thead>
<tr>
<th>State</th>
<th>Total, all lines</th>
<th>State</th>
<th>Total, all lines</th>
</tr>
</thead>
<tbody>
<tr>
<td>Alabama</td>
<td>$13,146,853</td>
<td>Montana</td>
<td>$3,807,820</td>
</tr>
<tr>
<td>Alaska</td>
<td>1,994,750</td>
<td>Nebraska</td>
<td>7,573,101</td>
</tr>
<tr>
<td>Arizona</td>
<td>17,959,495</td>
<td>Nevada</td>
<td>8,561,822</td>
</tr>
<tr>
<td>Arkansas</td>
<td>7,931,849</td>
<td>New Hampshire</td>
<td>3,342,091</td>
</tr>
<tr>
<td>California</td>
<td>111,243,502</td>
<td>New Jersey</td>
<td>29,106,640</td>
</tr>
<tr>
<td>Colorado</td>
<td>20,582,327</td>
<td>New Mexico</td>
<td>5,038,116</td>
</tr>
<tr>
<td>Connecticut</td>
<td>11,508,412</td>
<td>New York</td>
<td>61,861,773</td>
</tr>
<tr>
<td>D.C.</td>
<td>2,699,081</td>
<td>North Carolina</td>
<td>24,036,512</td>
</tr>
<tr>
<td>Delaware</td>
<td>3,842,583</td>
<td>North Dakota</td>
<td>3,826,625</td>
</tr>
<tr>
<td>Florida</td>
<td>88,067,510</td>
<td>Ohio</td>
<td>23,397,453</td>
</tr>
<tr>
<td>Georgia</td>
<td>32,039,933</td>
<td>Oklahoma</td>
<td>11,325,023</td>
</tr>
<tr>
<td>Hawaii</td>
<td>3,490,752</td>
<td>Oregon</td>
<td>10,510,197</td>
</tr>
<tr>
<td>Idaho</td>
<td>4,757,044</td>
<td>Pennsylvania</td>
<td>32,532,786</td>
</tr>
<tr>
<td>Illinois</td>
<td>35,938,656</td>
<td>Rhode Island</td>
<td>3,361,749</td>
</tr>
<tr>
<td>Indiana</td>
<td>16,321,562</td>
<td>South Carolina</td>
<td>15,078,257</td>
</tr>
<tr>
<td>Iowa</td>
<td>9,843,752</td>
<td>South Dakota</td>
<td>3,970,217</td>
</tr>
<tr>
<td>Kansas</td>
<td>9,565,831</td>
<td>Tennessee</td>
<td>17,442,243</td>
</tr>
<tr>
<td>Kentucky</td>
<td>10,147,626</td>
<td>Texas</td>
<td>92,331,270</td>
</tr>
<tr>
<td>Louisiana</td>
<td>16,226,096</td>
<td>Utah</td>
<td>8,100,467</td>
</tr>
<tr>
<td>Maine</td>
<td>3,316,037</td>
<td>Vermont</td>
<td>1,640,326</td>
</tr>
<tr>
<td>Maryland</td>
<td>16,186,978</td>
<td>Virginia</td>
<td>20,316,519</td>
</tr>
<tr>
<td>Massachusetts</td>
<td>21,453,245</td>
<td>Washington</td>
<td>18,109,893</td>
</tr>
<tr>
<td>Michigan</td>
<td>24,427,664</td>
<td>West Virginia</td>
<td>3,643,445</td>
</tr>
<tr>
<td>Minnesota</td>
<td>17,081,121</td>
<td>Wisconsin</td>
<td>14,623,479</td>
</tr>
<tr>
<td>Mississippi</td>
<td>7,505,441</td>
<td>Wyoming</td>
<td>1,820,520</td>
</tr>
<tr>
<td>Missouri</td>
<td>17,261,874</td>
<td><strong>United States</strong></td>
<td><strong>$949,898,320</strong></td>
</tr>
</tbody></table>

(1) Before reinsurance transactions, includes state funds, excludes territories.
(2) Data for the total United States may differ from similar data shown elsewhere due to the use of different exhibits from S&amp;P Global Market Intelligence.

Source: NAIC data, sourced from S&amp;P Global Market Intelligence, Insurance Information Institute.</td></tr>
</table></td> </tr>
      </tbody>
    </table>
  </body>
</html>