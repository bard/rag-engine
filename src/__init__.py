from . import ingest
from . import query
