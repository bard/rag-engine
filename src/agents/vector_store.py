from typing import List
from langchain.schema import Document
from langchain_pinecone import PineconeVectorStore
from langchain_openai import OpenAIEmbeddings
from pydantic import SecretStr


class VectorStore:
    def __init__(
        self,
        pinecone_api_key: SecretStr,
        pinecone_index_name: str,
        openai_api_key: SecretStr,
    ):
        """Initialize the vector store with OpenAI embeddings and Pinecone"""
        embeddings = OpenAIEmbeddings(api_key=openai_api_key)
        self.vector_store = PineconeVectorStore(
            index_name=pinecone_index_name,
            pinecone_api_key=pinecone_api_key.get_secret_value(),
            embedding=embeddings,
        )

    def add_documents(self, documents: List[Document]) -> None:
        """Add documents to the vector store"""
        self.vector_store.add_documents(documents)

    def similarity_search(self, query: str, k: int = 4) -> List[Document]:
        """
        Perform similarity search using the vector store.

        Args:
            query: Search query
            k: Number of results to return

        Returns:
            List of similar documents
        """
        return self.vector_store.similarity_search(query, k=k)
