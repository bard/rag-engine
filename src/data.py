from langchain.schema import Document
from typing import List
from pydantic import BaseModel
from bs4 import BeautifulSoup
from sqlalchemy.orm import Mapped, mapped_column, declarative_base


SqlAlchemyBase = declarative_base()


class InsuranceRecord(BaseModel):
    """Pydantic model for insurance record data"""

    year: int
    average_expenditure: float
    percent_change: float
    source_content: str


class SQLInsuranceRecord(SqlAlchemyBase):
    """SQLAlchemy model for insurance record data"""

    __tablename__ = "insurance_records"

    id: Mapped[int] = mapped_column(primary_key=True)
    year: Mapped[int] = mapped_column(nullable=False)
    average_expenditure: Mapped[float] = mapped_column(nullable=False)
    percent_change: Mapped[float] = mapped_column(nullable=False)
    source_content: Mapped[str] = mapped_column(nullable=False)

    def __repr__(self) -> str:
        return (
            f"SQLInsuranceRecord(id={self.id}, "
            f"year={self.year}, "
            f"average_expenditure={self.average_expenditure:.2f}, "
            f"percent_change={self.percent_change})"
        )


def parse_insurance_table(html_content: str) -> List[InsuranceRecord]:
    """
    Parse the auto insurance expenditure table and return a list of dictionaries.
    Each dictionary represents a row with year, expenditure, and percent change.
    """
    # Create BeautifulSoup object
    soup = BeautifulSoup(html_content, "html.parser")

    # Find the innermost table with the actual data
    tables = soup.find_all("table")
    data_table = tables[2]  # The third table contains our data

    # Initialize list to store results
    results = []

    # Process each row in tbody
    for row in data_table.tbody.find_all("tr"):
        cols = row.find_all("td")

        # Extract values
        year = int(cols[0].text.strip())

        # Clean and convert expenditure to float
        expenditure = cols[1].text.strip()
        expenditure = expenditure.replace("$", "").replace(",", "")
        expenditure = float(expenditure)

        # Clean and convert percent change to float
        percent_change = cols[2].text.strip()
        percent_change = percent_change.replace("%", "")
        percent_change = float(percent_change)

        # Create dictionary for current row
        record = InsuranceRecord(
            year=year,
            average_expenditure=expenditure,
            percent_change=percent_change,
            source_content=str(row),
        )
        results.append(record)

    return results


def insurance_record_to_langchain_document(
    record: InsuranceRecord, source_url: str
) -> Document:
    """
    Convert an InsuranceRecord to a LangChain Document.

    Args:
        record: The insurance record to convert
        source_url: The URL where the record was sourced from

    Returns:
        Document object containing the record data and metadata
    """
    content = "\n".join(
        [
            f"Year: {record.year}, "
            f"Expenditure: ${record.average_expenditure:.2f}, "
            f"Change: {record.percent_change}%"
        ]
    )

    return Document(
        page_content=content,
        id=f"insurance-record-{record.year}",
        metadata={
            "source_url": source_url,
            "source_content": record.source_content,
            "year": record.year,
            "average_expenditure": record.average_expenditure,
            "percent_change": record.percent_change,
        },
    )
