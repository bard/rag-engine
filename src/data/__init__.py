from .base import IndexableData
from .generic_tabular import GenericTabularData
from .insurance_average_expenditure import InsuranceAverageExpenditureData
from .textual import TextualData

__all__ = [
    "IndexableData",
    "GenericTabularData",
    "InsuranceAverageExpenditureData",
    "TextualData",
]
