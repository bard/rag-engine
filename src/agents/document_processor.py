from langchain.schema import Document
from sqlalchemy.orm import Session
from ..data import (
    SQLInsuranceRecord,
    insurance_record_to_langchain_document,
    parse_insurance_table,
)
from ..util import fetch_html_content


class DocumentProcessor:
    """A class for processing insurance data files into Document objects"""

    db: Session

    def __init__(self, db: Session):
        self.db = db

    def load_data(self, url: str) -> list[Document]:
        """
        Load insurance data from a URL and store it in the database.
        The data is sourced from the Insurance Information Institute (III)
        at https://www.iii.org/table-archive/20916

        Args:
            url (str): The URL to fetch insurance data from

        Returns:
            list[Document]: List of processed Document objects
        """

        html_content = fetch_html_content(url)
        insurance_records = parse_insurance_table(html_content)

        documents: list[Document] = []
        with self.db.begin():
            for r in insurance_records:
                self.db.add(SQLInsuranceRecord(**r.model_dump()))
                documents.append(
                    insurance_record_to_langchain_document(r, source_url=url)
                )

        return documents
