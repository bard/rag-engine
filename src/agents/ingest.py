import chromadb.utils.embedding_functions as ef
from langchain_community.embeddings.sentence_transformer import (
    SentenceTransformerEmbeddings,
)
from langchain_chroma import Chroma
from langchain_core.embeddings import Embeddings
from langchain_openai import OpenAIEmbeddings
from langchain.vectorstores import VectorStore
from langchain_core.runnables.config import RunnableConfig
from typing import TypedDict
from langchain.schema import Document
from pydantic import BaseModel, Field, SecretStr
from sqlalchemy import Engine, create_engine
from sqlalchemy.orm import Session
from ..data import (
    SQLInsuranceRecord,
    insurance_record_to_langchain_document,
    parse_insurance_table,
)
from ..util import fetch_html_content
from typing import Union, Literal


class PineconeVectorStoreBackend(BaseModel):
    type: Literal["pinecone"]
    index_name: str
    api_key: str
    embeddings_type: Literal["openai"]


class MockVectorStoreBackend(BaseModel):
    type: Literal["mock"]


class ChromaVectorStoreBackend(BaseModel):
    type: Literal["chroma"]
    path: str
    collection_name: str
    embeddings_type: Union[Literal["local-minilm"], Literal["openai"]]


class AgentConfig(BaseModel):
    openai_api_key: str
    db_url: str
    vector_store: Union[
        PineconeVectorStoreBackend, MockVectorStoreBackend, ChromaVectorStoreBackend
    ] = Field(discriminator="type")


class AgentState(TypedDict):
    url: str


def validate_agent_config(config: RunnableConfig) -> AgentConfig:
    """Validate agent config"""
    configurable = config.get("configurable")
    if configurable is None:
        raise Exception("Missing agent configuration")
    return AgentConfig(**configurable)


def ingest(state: AgentState, _config: RunnableConfig) -> AgentState:
    config = validate_agent_config(_config)

    url = state.get("url")
    if url is None:
        raise Exception("missing state")

    engine = create_engine(config.db_url)
    vector_store = get_vector_store(config)

    html_content = fetch_html_content(url)
    insurance_records = parse_insurance_table(html_content)
    documents: list[Document] = []

    with Session(engine) as session:
        with session.begin():
            for r in insurance_records:
                session.add(SQLInsuranceRecord(**r.model_dump()))
                documents.append(
                    insurance_record_to_langchain_document(r, source_url=url)
                )

    vector_store.add_documents(documents)

    return state


def get_vector_store(config: AgentConfig) -> VectorStore:
    if config.vector_store.type == "chroma":
        if config.vector_store.embeddings_type == "local-minilm":
            default_chroma_embedding_function = ef.DefaultEmbeddingFunction()
            if default_chroma_embedding_function is None:
                raise Exception("Default embedding function for Chroma is None")
            embedding_function = ChromaEmbeddingsAdapter(
                default_chroma_embedding_function
            )

        elif config.vector_store.embeddings_type == "openai":
            embedding_function = OpenAIEmbeddings(
                api_key=SecretStr(config.openai_api_key)
            )
        else:
            raise Exception(f"Unknown embedding type: {config.vector_store.type}")

        return Chroma(
            collection_name=config.vector_store.collection_name,
            embedding_function=embedding_function,
            persist_directory=config.vector_store.path,
        )
    else:
        raise Exception("boom")


def get_db(config: AgentConfig) -> Engine:
    return create_engine(config.db_url)


from langchain_core.embeddings import Embeddings
from chromadb.api.types import EmbeddingFunction


# https://cookbook.chromadb.dev/integrations/langchain/embeddings/#custom-adapter
class ChromaEmbeddingsAdapter(Embeddings):
    def __init__(self, ef: EmbeddingFunction):
        self.ef = ef

    def embed_documents(self, texts):  # pyright: ignore
        return self.ef(texts)

    def embed_query(self, query):  # pyright: ignore
        return self.ef([query])[0]
