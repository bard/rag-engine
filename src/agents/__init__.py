from .document_processor import DocumentProcessor
from .vector_store import VectorStore

__all__ = ["DocumentProcessor", "VectorStore"]
