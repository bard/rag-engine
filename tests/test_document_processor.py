from src.data import SQLInsuranceRecord
from src.agents.document_processor import DocumentProcessor


def test_document_processor_db_interactions(
    average_insurance_expenditures_html_as_data_url, mock_db
):
    # Create document processor instance
    processor = DocumentProcessor(mock_db)

    # Load data using the processor
    documents = processor.load_data(average_insurance_expenditures_html_as_data_url)

    # Verify database interactions
    assert mock_db.add.call_count == len(documents)
    mock_db.begin.assert_called_once()


def test_document_processor_return_value(
    average_insurance_expenditures_html_as_data_url, mock_db
):
    # Create document processor instance
    processor = DocumentProcessor(mock_db)

    # Load data using the processor
    documents = processor.load_data(average_insurance_expenditures_html_as_data_url)

    # Verify documents were returned
    assert len(documents) > 0
    assert all(hasattr(doc, "page_content") for doc in documents)
    assert all(
        doc.metadata["source_url"] == average_insurance_expenditures_html_as_data_url
        for doc in documents
    )
    assert documents[0].id == "insurance-record-2012"
    assert documents[0].metadata["year"] == 2012
    assert documents[0].metadata["average_expenditure"] == 812.40
    assert documents[0].metadata["percent_change"] == 2.2
    assert (
        documents[0].metadata["source_content"]
        == """<tr>
<td>2012</td>
<td>$812.40</td>
<td>2.2%</td>
</tr>"""
    )


def test_document_processor_database_state(
    average_insurance_expenditures_html_as_data_url, sqlite_session, snapshot
):
    processor = DocumentProcessor(sqlite_session)

    processor.load_data(average_insurance_expenditures_html_as_data_url)
    all_records = sqlite_session.query(SQLInsuranceRecord).all()

    assert all_records == snapshot
