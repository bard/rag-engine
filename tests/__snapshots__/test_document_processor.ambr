# serializer version: 1
# name: test_document_processor_database_state
  list([
    SQLInsuranceRecord(id=1, year=2012, average_expenditure=812.40, percent_change=2.2),
    SQLInsuranceRecord(id=2, year=2013, average_expenditure=841.06, percent_change=3.5),
    SQLInsuranceRecord(id=3, year=2014, average_expenditure=869.47, percent_change=3.4),
    SQLInsuranceRecord(id=4, year=2015, average_expenditure=896.66, percent_change=3.1),
    SQLInsuranceRecord(id=5, year=2016, average_expenditure=945.22, percent_change=5.4),
    SQLInsuranceRecord(id=6, year=2017, average_expenditure=1008.35, percent_change=6.7),
    SQLInsuranceRecord(id=7, year=2018, average_expenditure=1058.10, percent_change=4.9),
    SQLInsuranceRecord(id=8, year=2019, average_expenditure=1071.74, percent_change=1.3),
    SQLInsuranceRecord(id=9, year=2020, average_expenditure=1046.37, percent_change=-2.4),
    SQLInsuranceRecord(id=10, year=2021, average_expenditure=1061.54, percent_change=1.4),
  ])
# ---
